tools/rfstool

```diff
@@ -0,0 +1,475 @@
+# Make sure you have Docker installed and properly configured
+# (http://docker.com). Then,
+#   building: `./rfstool build`
+#     Then find the rootfs file under
+#     build/img/local/${HOSTTYPE}/minimal/fsfiles, where HOSTTYPE is the bash
+#     variable specifying your host architecture.
+#
+# Building is done inside a Docker container. Please make sure
+# you have Docker up and running on your system (see http:/docker.com) and your
+# user has permission to run Docker containers.
+#
+# The Firecracker sources dir will be bind-mounted inside the development
+# container (under /firecracker) and any files generated by the build process
+# will show up under the build/img/local/${HOSTTYPE} dir.
+#
+# By default, all rfstool commands run the container transparently, removing
+# it after the command completes. Any persisting files will be stored under
+# build/.
+#
+# Building:
+#   Run `./rfstool build`.
+#
+# Additional information:
+#   Run `./rfstool help`.
+#
+# TODO:
+#   - minimal_with_initrd build
+#   - ubuntu build
+#   - ubuntu_with_ssh build
+#   - all build
+
+# Rootfs build container image (name:tag)
+# This should be updated whenever we upgrade the rootfs build container.
+RFSCTR_IMAGE="fcuvm/rfs:v01"
+
+# Naming things is hard
+MY_NAME="Firecracker $(basename "$0")"
+
+# Full path to the Firecracker tools dir on the host.
+FC_TOOLS_DIR=$(cd "$(dirname "$0")" && pwd)
+
+# Full path to the Firecracker sources dir on the host.
+FC_ROOT_DIR=$(cd "${FC_TOOLS_DIR}/.." && pwd)
+
+# Full path to the build dir on the host.
+FC_BUILD_DIR="${FC_ROOT_DIR}/build"
+
+# Full path to the img dir on the host.
+FC_IMAGES_DIR="${FC_BUILD_DIR}/img"
+
+# Full path to the Firecracker sources dir, as bind-mounted in the container.
+CTR_FC_ROOT_DIR="/firecracker"
+
+# Full path to the build dir, as bind-mounted in the container.
+CTR_FC_BUILD_DIR="${CTR_FC_ROOT_DIR}/build"
+
+# Full path to the microVM images cache dir
+CTR_MICROVM_IMAGES_DIR="${CTR_FC_BUILD_DIR}/img"
+
+# Rootfs Alpine version
+RFS_ALPINE_VERS="3.8.1"
+
+# Alpine CDN url
+RFS_ALPINE_ORIGIN="http://dl-cdn.alpinelinux.org"
+
+# Rootfs Alpine x86_64 package name, url locations, and sha256.
+RFS_ALPINE_X86_64_GZ="alpine-minirootfs-${RFS_ALPINE_VERS}-x86_64.tar.gz"
+RFS_ALPINE_X86_64_URL="${RFS_ALPINE_ORIGIN}"
+RFS_ALPINE_X86_64_URL+="/alpine/v${RFS_ALPINE_VERS%.*}/releases/x86_64/"
+RFS_ALPINE_X86_64_URL+="/${RFS_ALPINE_X86_64_GZ}"
+RFS_ALPINE_X86_64_SHA256="ad753d802048fa902e4d8b35cc53656de8ed0e6d082246089a"
+RFS_ALPINE_X86_64_SHA256+="11a86014b0f1a5"
+
+# Rootfs Alpine aarch64 package name, url location, and sha256.
+RFS_ALPINE_AARCH64_GZ="alpine-minirootfs-${RFS_ALPINE_VERS}-aarch64.tar.gz"
+RFS_ALPINE_AARCH64_URL="${RFS_ALPINE_ORIGIN}"
+RFS_ALPINE_AARCH64_URL+="/alpine/v${RFS_ALPINE_VERS%.*}/releases/aarch64/"
+RFS_ALPINE_AARCH64_URL+="/${RFS_ALPINE_AARCH64_GZ}"
+RFS_ALPINE_AARCH64_SHA256="425be487ecc4bd0333fd84ec8c6c4b971a32446be76df26e7"
+RFS_ALPINE_AARCH64_SHA256+="add04be95987fa8"
```

Comment:
Can we store this in an associative array or some sort of key-value object for readability ?  Again I'm not an expert in bash so I'm not sure about this. Ideally we could have something like (please ignore the syntax. It's inspired by javascript, but I hope we can do something like this in bash):
```
images = {
  minimal: {
    x86_64: {
      gz: ...
      url: ....
      sha: ....
    }
  }
}
```

This way we can then have some generic logic for downloading the image. If you know the image name (e.g. minimal`) and architecture (e.g. x86_64). You can just look in the key-value store and find all the parametrs that you need in order to download the sources.

Reply:
If the Alpine version is expected to be the same on both architectures,  then I think this just becomes:
```
RFS_ALPINE_GZ="alpine-minirootfs-${RFS_ALPINE_VERS}-${HOSTTYPE}.tar.gz"
RFS_ALPINE_URL="${RFS_ALPINE_ORIGIN}"
RFS_ALPINE_URL+="/alpine/v${RFS_ALPINE_VERS%.*}/releases/${HOSTTYPE}/"
RFS_ALPINE_URL+="/${RFS_ALPINE_GZ}"
```
And GZ and URL don't need to be included in any associative array,  just the SHA value.  I think.  

Currently however, the Alpine version for "minimal" differs between architecture, so the rootfs stored in S3 seem to me to be 3.8 on x86_64, 3.9 on aarch64.  So I guess it depends on the intention.  If it's to track each rootfs version per architecture,  i.e. try and match the discrepancies (which I appreciate could be intentional), then yes,  I will put in an array as you suggest.  If it's to clobber anything that's currently there and keep test artefacts aligned between architectures, then perhaps I don't need to.

Reply:
Also,  I'm not sure that bash is very good at associative arrays, from what I've been reading,  bash's associative arrays are "one-dimensional", i.e. don't nest well.   Unless I've just not found the right  way....